compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2056,13 +2056,37 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                 }
             }
             ObligationCauseCode::ImplDerivedObligation(ref data) => {
-                let parent_trait_ref = self.resolve_vars_if_possible(&data.parent_trait_ref);
+                let mut parent_trait_ref = self.resolve_vars_if_possible(&data.parent_trait_ref);
                 err.note(&format!(
                     "required because of the requirements on the impl of `{}` for `{}`",
                     parent_trait_ref.print_only_trait_path(),
                     parent_trait_ref.skip_binder().self_ty()
                 ));
-                let parent_predicate = parent_trait_ref.without_const().to_predicate(tcx);
+
+                let mut parent_predicate = parent_trait_ref.without_const().to_predicate(tcx);
+                let mut data = data;
+                let mut redundant = false;
+                let mut count = 0;
```

Comment:
While this PR addresses the direct self-recursion case, it doesn not handle mutual recursion at all (example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0d09d2ecaa6ddc83765340d9afa1e7ee). In order to solve this you're going to have to create a hashset of `DefId`s to see if an id showed up before

I think it should also be possible to emit the full diagnostic on demand. Do we have a `verbose` flag we can use for this? `-Zverbose` is only available on nightly.

Reply:
I can easily add a check for `-Zverbose`, but I do think that most of the time this info is not useful at all.